File: leetcode/1160/solution.go
/**
 * https://leetcode.com/problems/find-words-that-can-be-formed-by-characters/
 * 1160. Find Words That Can Be Formed by Characters
 */
package main

func countCharacters(words []string, chars string) int {
	base := toCounter(chars)
	total := 0
	for _, w := range words {
		flag := true
		for i, cnt := range toCounter(w) {
			if cnt > base[i] {
				flag = false
				break
			}
		}
		if flag {
			total += len(w)
		}
	}
	return total
}

func toCounter(s string) [26]int {
	arr := [26]int{}
	for _, b := range []byte(s) {
		i := b - 'a'
		arr[i] += 1
	}
	return arr
}

File: leetcode/1160/solution.rs
/**
 * https://leetcode.com/problems/find-words-that-can-be-formed-by-characters/
 * 1160. Find Words That Can Be Formed by Characters
 */
impl Solution {
    pub fn count_characters(words: Vec<String>, chars: String) -> i32 {
        let base = to_arr(&chars);
        let mut total = 0;
        for word in words {
            let mut can_form = true;
            for (&count, base_count) in to_arr(&word).iter().zip(base) {
                if count > base_count {
                    can_form = false;
                    break;
                }
            }
            if can_form {
                total += word.len();
            }
        }

        total as i32
    }
}

fn to_arr(s: &str) -> [i32; 26] {
    let mut arr = [0; 26];
    for b in s.bytes() {
        let i = (b - b'a') as usize;
        arr[i] += 1;
    }
    arr
}
